### DISCUSSION TITLE
[3일차 - 예진] 스코프

### DISCUSSION THREAD

**Discussion by ez1n** [zzangCS/zzangInterview#13] [프론트 면접 질문]
스코프는 무엇인가요?

**Comment by ez1n** [comment]
**스코프** 
- 변수가 접근할 수 있는 범위
- 한 스코프 범위 내에서 외부 스코프의 변수에는 접근 가능하지만 내부 스코프의 변수에는 접근이 불가능함

1. 지역 스코프
- 한정적인 범위에서 해당 변수에 접근이 가능함 (해당 지역을 벗어나면 저근 불가능)
- 함수 스코프, 블록 스코프

2. 전역 스코프
- 어디에서든 해당 변수에 접근이 가능함

**Reply by ez1n** [reply]
- **스코프 (scope)**
> 변수가 값을 참조할 때 접근할 수 있는 범위


- **전역 스코프 (global scope)**

> 전역에 선언되어 있어 어느 곳에든지 해당 변수에 접근할 수 있음 (전역 변수)

- **지역 스코프 (local scope)**

> 해당 지역에서만 접근할 수 있기 때문에 지역을 벗어난 곳에서는 접근이 불가능함 (지역 변수)

⇒ 함수 스코프 : 함수에서 선언한 변수는 해당 함수 내에서만 접근이 가능함
⇒ 블록 스코프 : 블록 내부에서 선언된 변수는 해당 블록에서만 접근이 가능함
    
- **Lexical Scope vs Dynamic Scope**

⇒ Lexical Scope : 함수 선언 위치에 따라 결정되는 상위 스코프
⇒ Dynamic Scope : 함수 호출 위치에 따라 결정되는 상위 스코프

```
💡 전역 변수는 최소화하는 것이 좋다.

전역 변수를 최소화하면 side effect를 줄일 수 있다.
```

**Reply by ez1n** [reply]
https://medium.com/@su_bak/javascript-%EC%8A%A4%EC%BD%94%ED%94%84-scope-%EB%9E%80-bc761cba1023

**Comment by smarfy99** [comment]
변수에 접근할 수 있는 범위로 global과 local이 있다.
전역 스코프는 전역으로 선언되어 어느 곳에서든지 해당 변수에 접근할 수 있고, 지역 스코프는 해당 범위에서만 접근할 수 있고, 범위를 벗어난 곳에서는 접근할 수 없다.

**Reply by smarfy99** [reply]
https://medium.com/@su_bak/javascript-%EC%8A%A4%EC%BD%94%ED%94%84-scope-%EB%9E%80-bc761cba1023

**Comment by skyhanull** [comment]
함수와 함수가 선언된 어휘적 환경의 조합입니다. 또한 **함수가 속한 렉시컬스코프를 기억하여 함수가 렉시컬 스코프 밖에서 실행될 때도 그 스코프에 접근할 수 있게 하는 기능**
 을 말합니다.
